src/control/crawler.c

```diff
@@ -840,6 +841,204 @@ void dt_control_crawler_show_image_list(GList *images)
                    G_CALLBACK(dt_control_crawler_response_callback), gui);
 }
 
+/* thumbs updating in backthread stuff
+  1. timing stuff has been made user configurable
+  2. database timestamps are kept via conf to exclude most data while scanning the
+      database for less file testing
+*/
+
+static int64_t _get_timestamp(const char *key)
+{
+  const char *str = dt_conf_get_string_const(key);
+  int64_t res = 0;
+  if(str)
+    sscanf(str, "%" SCNi64, &res);
+  return res;
+}
+
+static void _set_timestamp(const char *key, int64_t val)
+{
+  char str[128] = { 0 };
+  g_snprintf(str, sizeof(str), "%" PRIi64, val);
+  dt_conf_set_string(key, str);
+}
+
+static dt_mipmap_size_t _get_mipsize(void)
+{
+  const char *option = dt_conf_get_string_const("backthumbs_mipsize");
+  const dt_mipmap_size_t m = dt_mipmap_cache_get_min_mip_from_pref(option);
+  return (m >= DT_MIPMAP_1 && m <= DT_MIPMAP_7) ? m : 0;
+}
+
+static gboolean _mipfile_existing(const dt_imgid_t imgid, const dt_mipmap_size_t k)
+{
+  char filename[PATH_MAX] = { 0 };
+  snprintf(filename, sizeof(filename), "%s.d/%d/%d.jpg", darktable.mipmap_cache->cachedir, k, imgid);
+  return dt_util_test_image_file(filename);
+}
+
+static inline gboolean _still_thumbing(const float inactivity)
+{
+  return dt_control_running() && dt_lighttable_silent(inactivity);
+}
+
+static void _update_img_thumbs(const dt_imgid_t imgid, const dt_mipmap_size_t max_mip)
+{
+  for(int k = max_mip; k >= DT_MIPMAP_1; k--)
+  {
+    // if a valid thumbnail file is already on disc - do nothing
```

Comment:
Not sure to understand, what if the thumbnail is out of date?

Reply:
From what I know, all mipmaps are removed if invalid if there are changes so that seemed safe to me.

Will check if that could lead to race conditions so it might be better to enforce writing new mipmaps.

Reply:
done, indeed might be necessary